File: backend/models/graph.py
from uuid import uuid4

from sqlalchemy import (
    Boolean,
    Column,
    Float,
    ForeignKey,
    Integer,
    JSON,
    String,
)
from sqlalchemy.orm import relationship

from backend.database import Base


class Site(Base):
    __tablename__ = "sites"

    id = Column(String, primary_key=True, default=lambda: str(uuid4()))
    name = Column(String, nullable=False)
    role = Column(String, default="spoke")  # hub | spoke
    wan_type = Column(String, default="mpls_aviat")  # mpls_aviat | sdwan
    observable_boundary = Column(String, nullable=True)
    canvas_x = Column(Float, default=0.0)
    canvas_y = Column(Float, default=0.0)
    canvas_w = Column(Float, default=400.0)
    canvas_h = Column(Float, default=300.0)

    nodes = relationship(
        "NetworkNode", back_populates="site", cascade="all, delete-orphan"
    )


class NetworkNode(Base):
    __tablename__ = "nodes"

    id = Column(String, primary_key=True, default=lambda: str(uuid4()))
    site_id = Column(String, ForeignKey("sites.id"), nullable=True)
    label = Column(String, nullable=False)
    # core_internal | core_external | aviat_ctr | sdwan_cpe | access_switch
    node_type = Column(String, nullable=False)
    vendor = Column(String, default="Cisco")
    management_ip = Column(String, nullable=True)
    role = Column(String, nullable=True)
    zone = Column(String, nullable=True)
    observable = Column(Boolean, default=True)
    wan_facing = Column(Boolean, default=False)
    signal_dbm = Column(Float, nullable=True)
    port_count = Column(Integer, nullable=True)
    meta = Column(JSON, default=dict)
    position_x = Column(Float, default=0.0)
    position_y = Column(Float, default=0.0)

    site = relationship("Site", back_populates="nodes")

    source_edges = relationship(
        "NetworkEdge",
        foreign_keys="[NetworkEdge.source_id]",
        back_populates="source_node",
        cascade="all, delete-orphan",
    )
    target_edges = relationship(
        "NetworkEdge",
        foreign_keys="[NetworkEdge.target_id]",
        back_populates="target_node",
        cascade="all, delete-orphan",
    )


class NetworkEdge(Base):
    __tablename__ = "edges"

    id = Column(String, primary_key=True, default=lambda: str(uuid4()))
    source_id = Column(String, ForeignKey("nodes.id"), nullable=False)
    target_id = Column(String, ForeignKey("nodes.id"), nullable=False)
    edge_type = Column(String, default="fiber")  # fiber | mpls | sdwan
    vrf = Column(String, nullable=True)
    capacity_mbps = Column(Float, nullable=True)
    meta = Column(JSON, default=dict)

    source_node = relationship(
        "NetworkNode",
        foreign_keys=[source_id],
        back_populates="source_edges",
    )
    target_node = relationship(
        "NetworkNode",
        foreign_keys=[target_id],
        back_populates="target_edges",
    )

File: frontend/src/api/graph.ts
import apiClient from './client';
import type { Node, Edge } from 'reactflow';
import type { TopologyGraphSchema } from '../types/api';

// ── Payload builder (pure) ────────────────────────────────────────────────────
// Shared by GraphCanvas (manual save) and PropertiesPanel (auto-save on delete).

export function buildSavePayload(nodes: Node[], edges: Edge[]): TopologyGraphSchema {
  const siteNodes = nodes.filter((n) => n.type === 'siteGroup');
  const regularNodes = nodes.filter((n) => n.type !== 'siteGroup');
  return {
    sites: siteNodes.map((n) => ({
      id: n.id,
      name: n.data?.label ?? n.id,
      role: n.data?.role ?? 'spoke',
      wan_type: n.data?.wan_type ?? 'mpls_aviat',
      observable_boundary: n.data?.observable_boundary ?? null,
      canvas_x: n.position.x,
      canvas_y: n.position.y,
      canvas_w: (n.style?.width as number) ?? 400,
      canvas_h: (n.style?.height as number) ?? 300,
    })),
    nodes: regularNodes.map((n) => ({
      id: n.id,
      site_id: (n as any).parentNode ?? n.data?.siteId ?? '',
      label: n.data?.label ?? n.id,
      node_type: n.type ?? 'access_switch',
      vendor: n.data?.vendor ?? 'Cisco',
      management_ip: n.data?.management_ip ?? null,
      role: n.data?.role ?? null,
      zone: n.data?.zone ?? null,
      observable: n.data?.observable ?? true,
      position_x: n.position.x,
      position_y: n.position.y,
      meta: {},
    })),
    edges: edges.map((e) => ({
      id: e.id,
      source_id: e.source,
      target_id: e.target,
      edge_type: e.type ?? 'fiber',
      vrf: e.data?.vrf ?? null,
      capacity_mbps: e.data?.capacity_mbps ?? null,
    })),
  };
}

// ── API calls ─────────────────────────────────────────────────────────────────

export const getGraph = async (): Promise<TopologyGraphSchema> => {
  const res = await apiClient.get<TopologyGraphSchema>('/api/graph/');
  return res.data;
};

export const saveGraph = async (graph: TopologyGraphSchema): Promise<void> => {
  await apiClient.post('/api/graph/', graph);
};

export const updateNode = async (nodeId: string, data: Partial<unknown>): Promise<void> => {
  await apiClient.put(`/api/graph/nodes/${nodeId}`, data);
};

export const deleteNode = async (nodeId: string): Promise<void> => {
  await apiClient.delete(`/api/graph/nodes/${nodeId}`);
};

export const exportGraph = async (): Promise<unknown> => {
  const res = await apiClient.get('/api/graph/export');
  return res.data;
};
